Add Websocket Support to pepper.js with testing infrastructure

danielrh opened the pull request:
Right now pepper.js doesn't support the websocket interfaces that ppapi presents.
This implements the interfaces and provides a testing suite to test them.

danielrh committed:
WebSocket pepper.js implementation with included websocket test

danielrh commented:
I tested this under pnacl debug, emscripten debug and release...
however under pnacl release I get this error (probably easy to solve)

```
make[2]: Leaving directory `/home/daniel/dev/pepper.js/examples/ppapi'
  CXX  pnacl/Release/ppapi_websocket.o
  CXX  pnacl/Release/web_socket_request.o
  CXX  pnacl/Release/module.o
  LINK pnacl/Release/ppapi_websocket.bc
  FINALIZE pnacl/Release/ppapi_websocket.pexe
make[1]: *** No rule to make target `pnacl/Release/ppapi_websocket_unstripped.bc', needed by `pnacl/Release/ppapi_websocket.nmf'.  Stop.
make[1]: Leaving directory `/home/daniel/dev/pepper.js/examples/tests/ppapi_websocket'
```

any idea how to solve that one?

danielrh commented:
oh silly me: the makefile rules updated
This fixes it
https://github.com/danielrh/pepper.js/commit/7298f642a5259420ff0420f4f8ff497b2f9fb519

Now all websockets tests pass on pnacl Debug and Release and emscripten Debug and Release

ncbray commented:
(At JSConf.  I'll give this a look when I have a moment, but that may not be until Monday.)

danielrh commented:
Thanks: will much appreciate it when you get a chance :-)

ncbray commented on examples/tests/ppapi_websocket/module.cc:
```diff
@@ -0,0 +1,329 @@
+
+}  // namespace
+
+PP_Bool Instance_DidCreate(PP_Instance /*pp_instance*/,
```
Be consistent with rest of code: do not comment out unused parameter names.

danielrh replied:
Will fix: copied that style from ppapi_geturl/module.cc

ncbray commented on tools/nacl_emscripten.mk:
```diff
@@ -46,9 +46,9 @@ NACL_LDFLAGS+=-Wno-warn-absolute-paths
 # It appears that emcc currently assumes --minify will always be used to disable minification, so --minify 1 does not work, rather it also disables minification.
 ifeq ($(CONFIG),Release)
-NACL_LDFLAGS+=-O2 -s RESERVED_FUNCTION_POINTERS=325 --closure 1
+NACL_LDFLAGS+=-O2 -s RESERVED_FUNCTION_POINTERS=512 --closure 1
```
Please tune this to a smaller number.  (400?)  Injecting function pointers causes problems for asm (it reserves X function pointers for each possible function type, so it's inefficeint).  Down the road I'll need to rework the system to avoid using this parameter, but for the meantime it would be nice not to wildly overallocate the number of function pointers.

You should probably also add a comment to this effect, although I'd also be happy to do it myself later (I didn't realize this should be documented until I saw your change).

ncbray commented on ppapi_preamble.js:
```diff
@@ -94,6 +94,8 @@ var GRAPHICS_3D_RESOURCE = 20;
 var ARRAY_RESOURCE = 21;
 var DICTIONARY_RESOURCE = 22;
+var WEB_SOCKET_RESOURCE = 23;
+
```
Do not add extra blank line.

ncbray commented on wrappers/web_socket.js:
```diff
@@ -0,0 +1,303 @@
+// Copyright (c) 2013 Google Inc. All rights reserved.
```
For all new files that aren't modified copies, the copyright should be 2014.
The attributions for anything you created should be pepper.js Authors or Chromium Authors.  Add yourself to CONTIBUTORS.  Add yourself to AUTHORS if you haven't signed a Chromium contributor's agreement.

ncbray commented on wrappers/web_socket.js:
```diff
@@ -0,0 +1,303 @@
+    if (protocol_count == 1) {
+        protocol = [glue.memoryToJSVar(protocols)];
+    }
+    socket = resources.resolve(socket, WEB_SOCKET_RESOURCE);
```
Create a new local instead of reusing?  Keeping the code implicitly statically typed makes the JITs job easier.

ncbray commented on wrappers/web_socket.js:
```diff
@@ -0,0 +1,303 @@
+    url = glue.memoryToJSVar(url);
+    callback = glue.getCompletionCallback(callback);
+    var req;
+    req = new WebSocket(url, protocol); // FIXME: deal with protocol array
```
Merge with previous line?

FIXME => TODO(danielrh)   This doesn't put you on the hook, it just documents who added the TODO, if there are any questions.

ncbray commented on wrappers/web_socket.js:
```diff
@@ -0,0 +1,303 @@
+    if (socket === undefined) {
+      return ppapi.PP_VARTYPE_UNDEFINED;
+    }
+    if (!socket.websocket){
```
Comment when this can occur... or just make an isConnected helper method to make this check self documenting?

ncbray commented on wrappers/web_socket.js:
```diff
@@ -0,0 +1,303 @@
+  var WebSocket_IsWebSocket = function(resource) {
+    return resources.is(resource, WEB_SOCKET_RESOURCE);
+  };
+  function performCallback(socket, data){
```
Dead code.
(Not resolved, yet.)

ncbray commented on examples/tests/ppapi_websocket/web_socket_request.cc:
```diff
@@ -0,0 +1,239 @@
+#include <string>
+#include <sstream>
+
+//#include "native_client/src/include/portability.h"
```
No dead code unless you have a very good (and commented) reason.

ncbray commented on wrappers/web_socket.js:
```diff
@@ -0,0 +1,303 @@
+  var WebSocket_Connect = function(socket, url, protocols, protocol_count, callback) {
+    if (protocol_count > 1) {
+        // currently don't support protocol lists
+        return ppapi.PP_ERROR_FAILED;
```
Not 100% sure, but it may be better to throw an exception when you hit implementation flaws like this.  Inform the developer.

ncbray commented on wrappers/web_socket.js:
```diff
@@ -0,0 +1,303 @@
+    if (socket === undefined) {
+      return ppapi.PP_ERROR_BADRESOURCE;
+    }
+    url = glue.memoryToJSVar(url);
```
Var reuse, again.  Will no longer comment on it.

ncbray commented on wrappers/web_socket.js:
```diff
@@ -0,0 +1,303 @@
+    socket.receiveQueue = new Queue();
+
+    req.onclose = function (evt) {
+        if (req.closeCallback !== null && !socket.dead) {
```
{req => socket}.closeCallback?

I wouldn't be against changing from httpd.py to a simple node webserver if you want to write tests.  This change is already big enough, however, that that's a discussion for later.

ncbray commented on wrappers/web_socket.js:
```diff
@@ -0,0 +1,303 @@
+    if (reason.length > 123) {
+        return ppapi.PP_ERROR_BADARGUMENT;
+    }
+    if (code_u16 != 1000 && !(code_uint16 >= 3000 && code_uint16 <4999)) {
```
space before 4999

ncbray commented on examples/tests/ppapi_websocket/ppapi_websocket.html:
```diff
@@ -0,0 +1,94 @@
+        addTest('Valid', VALID_URL, '6162636465666768696a6b6c6d6e6f70', '6162636465666768696a6b6c6d6e6f70', true);//abcdefghijklmnop
+        addTest('Valid_File', VALID_URL, 'f0288c28000a0dffeeeffe', 'f0288c28000a0dffeeeffe', true);//invalid unicode codepoint
+        addTest('Invalid', INVALID_URL, "WebSocketRequest::OpenCallback: PP_ERROR_FAILED", "WebSocketRequest::OpenCallback: PP_ERROR_FAILED", false);
+//        addTest('Invalid_File', INVALID_URL, '0a0a0a0a', '404', false);
```
Dead code.  Document intent or remove.

ncbray commented:
First pass at reviewing.  Mostly small mechanical stuff, will take a second, higher-level pass afterwards.

danielrh commented:
Ok applied your comments :-)

ncbray commented on wrappers/web_socket.js:
```diff
@@ -0,0 +1,305 @@
+  var WebSocket_GetURL = function(socketResource) {
+    var socket = resources.resolve(socketResource, WEB_SOCKET_RESOURCE);
+    if (socket === undefined) {
+      return ppapi.PP_VARTYPE_UNDEFINED;
```
I didn't have time to start reviewing this until late in the day, but I thought I'd send out a quick note: I suspect many of your return values are wrong.  In this case, when the function must return a struct, Emscripten transforms the function by adding an extra argument that points to where the return value should be stored.

See: https://github.com/google/pepper.js/blob/master/wrappers/base.js#L130

vs.

https://developer.chrome.com/native-client/pepper_stable/c/struct_p_p_b___var__1__2#a5c7ce303ae72998f54ba5da3357f7641

Any time you're returning a PP_Var by value, you will need to set the type and value fields of the result pointer.  (undefined or null PP_Var only need the type set, however.)  asm.js also doesn't have a string type, so the next return statement is also wrong - you'd need to return a PP_Var of type string containing an empty string value.  This will be a simplified version of the function I linked to, specialized for a specific value, and you probably want to put that in its own helper function.

Yes... all this poking at pointers is a pain in the butt.  :/  Ask if you have questions.

ncbray commented:
See the newest comment, I haven't had a lot of time to review, but I suspect return values need some love.

danielrh commented:
Good catch! I modified the code to properly return strings using the 0th parameter of the appropriate javascript functions.
I added testing to ensure that the url was correctly computed: apparently we need a slash for it to be properly canonicalized, so that was a bonus catch from adding the GetURL test.

danielrh commented:
Any chance to review the new diff?

ncbray commented on wrappers/web_socket.js:
```diff
@@ -0,0 +1,328 @@
+    }
+    return socket.closeWasClean;
+  };
+  var WebSocket_GetExtensions = function(result, socketResource) {
```
Empty line between functions.

ncbray commented on wrappers/web_socket.js:
```diff
@@ -0,0 +1,328 @@
+  var returnString = function(result, string){
+    var memory = 0;
+    var len = string.length;
+    if (len > 0) {
```
Treat the empty string like any other string.  A slight inefficiency, yes, but this only happens in the error case.  Doing so guarentees the data pointer will always be valid and removes one extra thing to think about.

ncbray commented on wrappers/web_socket.js:
```diff
@@ -0,0 +1,328 @@
+  };
+
+  var WebSocket_Close = function(socket, codeUint16, reason, callback) {
+    socket = resources.resolve(socket, WEB_SOCKET_RESOURCE);
```
Not updated to socketResource.

danielrh replied:
This is an internal function operating on the raw socket, not the Resource... won't be called from outside

ncbray replied:
To clarify, you're changing the type of "socket".  You modified the other functions to avoid this by changing the name of the parameter to "socketResource".  You missed this function in an earlier cleanup.

ncbray commented on wrappers/web_socket.js:
```diff
@@ -0,0 +1,328 @@
+    if (!hasSocketClassConnected(socket)) {
+        return ppapi.PP_ERROR_FAILED;
+    }
+    reason = glue.memoryToJSVar(reason);
```
Dito.

ncbray commented on wrappers/web_socket.js:
```diff
@@ -0,0 +1,328 @@
+
+    return ppapi.PP_OK_COMPLETIONPENDING;
+  };
+  var hasSocketClassConnected = function(socket) {
```
Empty line between functions.

ncbray commented on wrappers/web_socket.js:
```diff
@@ -0,0 +1,328 @@
+      return ppapi.PP_ERROR_BADRESOURCE;
+    }
+    var callback = glue.getCompletionCallback(cCallback);
+    var req;
```
Merge "var" decl and assignment for consistency.

ncbray commented on wrappers/web_socket.js:
```diff
@@ -0,0 +1,328 @@
+            socket.closeCallback(ppapi.PP_OK);
+            socket.closeCallback = null;
+        }
+        socket.closeWasClean = evt.wasClean;
```
This function is correct but looks a little suspicious unless you know the callback is deferred until the next turn of the event loop.  Swapping the assignment block and the if statement would avoid the need to thinking or a comment explaining why you don't have to think. :)

ncbray commented on examples/tests/ppapi_websocket/web_socket_request.h:
```diff
@@ -0,0 +1,67 @@
+#include "ppapi/c/pp_instance.h"
+#include "ppapi/c/pp_resource.h"
+#include "ppapi/c/pp_stdint.h"
+//#include "ppapi/c/trusted/ppb_file_io_trusted.h"
```
No dead code.

ncbray commented on wrappers/web_socket.js:
```diff
@@ -0,0 +1,328 @@
+        returnString(result, '');
+        return;
+    }
+
```
Remove empty line for consistency.

ncbray commented:
OK, all I've found are embarrassingly pedantic style issues and a few opportunities for defensive programming.  Once you take care of these, I'll be happy to merge.

FYI I'll be squashing the branch into a single commit on the master branch.

danielrh commented:
Ok done :-)

danielrh commented:
Squashed

ncbray commented:
Sorry about the lag.  As an act of penance, I did two changes you missed so I could push it ASAP.  (The commit on master will be slightly different than your branch.)

danielrh commented:
Nice :-D will rebase to yours! thanks so much
